### PR TITLE
Improve display of reasons an NPC won't trade an item

### LINKED ISSUE #61597: Unable to trim string 'Hub 01 will not buy back their own gear' to width 0.  Returning empty string.
### Describe the bug

Got this error:
```
 DEBUG    : Unable to trim string 'Hub 01 will not buy back their own gear' to width 0.  Returning empty string.

 FUNCTION : trim_by_length
 FILE     : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\output.cpp
 LINE     : 275
 VERSION  : 737dc71
```
When I was trading with hub01 and scrolled down to the next page of my inventory to trade

### Steps to reproduce

1. Go to hub01 trader
2. tab to your inventory selection
3. scroll down to the next page

### Expected behavior

Scroll to the next page without showing the error

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.22000.978 (21H2)
- Game Version: 737dc71 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions],
    Magiclysm [magiclysm],
    Tamable Wildlife [Tamable_Wildlife]
]

### Additional context

Possibly related to #61591?

### PR DESCRIPTION
**Pull request by ZeroInternalReflection**
#### Summary
Interface "Improve display of reasons an NPC won't trade an item"

#### Purpose of change
If an NPC is unwilling to buy or sell an item, the player should have some idea why.  The Trade UI tries to fit this in, but there are a couple of issues with how it does so:
1] For long item names, the trade denial message is shrunk to fit the _untrimmed_ item name.  Note the trimming on the "duffel bag" line.  When the screen is two characters narrower, it will produce an error message as in #61597.
Before--Improper_trimming

Fixes #61597 

2] There's no way to scroll through the entry, so on small screen sizes, the full denial message is unavailable.  So, the player might not know why they can't buy/sell an item
3] On languages with wide characters, there's a minor display issue where the denial message is shifted to the left:
Before--Improper_positioning--Chinese

#### Describe the solution
1] Trim the denial message to fit with the trimmed item name, ensuring there's always at least something indicating that this trade is unavailable.
After--Proper_trimming
2] When attempting to select an untradeable item for trade, display a popup with the full item name and the reason it cannot be traded
After--Denial_popup--English
3] Change form trim_and_print() to right_print(trim_by_length()) to right-justify the denial message, even on languages with wide characters

#### Describe alternatives you've considered
1. Adding the ability to scroll through the text in a given inventory_selector line
This seemed like it would be a nightmare to control

#### Testing
Tested at a variety of screen sizes:

https://github.com/CleverRaven/Cataclysm-DDA/assets/89038572/30bda7af-6c6f-4bdf-af43-d311c4b035c7

https://github.com/CleverRaven/Cataclysm-DDA/assets/89038572/cfa2a7f7-423e-4635-bd04-73c5271409e9

Tested in several different languages:
After--Denial_popup--Spanish

https://github.com/CleverRaven/Cataclysm-DDA/assets/89038572/aa59114d-94a0-40bf-ac82-28aac660ef5e

Tested trading with Hub 01:
After--Hub01_trading

Tested pickup screen denial popup (Portuguese shown):
After--Pickup_screen_denial--Portuguese

Tested wear item/disassembly screens to ensure I didn't break display of denial messages on other inventory_selector screens.
If there's a way to get a denial message on comparison `I` or drop `D`, that should probably be tested, but I didn't see a way to do it.

#### Additional context
- It might be useful to add the popup denial message to other inventory_selector types.  In some languages it doesn't seem like there's a lot of room to describe, e.g., the tools required for disassembly:
After--Disassembly--Polish

### CODE REVIEW

**Comment by Zireael07** [comment]
Hm... another, fairly obvious thing that jumps out to me is that we could move the merchant's name to the screen header (where TRADE is) therefore making the denial message itself more visible instead of X repetitions of <merchant's name> <something cut off>

**Comment by ZeroInternalReflection** [comment]
> we could move the merchant's name to the screen header (where TRADE is)
I'm in favour of cutting down on the repetition in some of the denial messages, but the merchant's name actually already appears in the header (bottom left of the top pane).  On certain languages/screen sizes, the name does get cut off by the keybinding hint for autobalancing, though, so that's not ideal.

I think the to-do list would be:
1. Tweak the header so information remains visible at small sizes
2. Change a bunch of the "[Merchant's name] will never trade this" to a less repetitive and more descriptive message.  Perhaps "Will not trade personal items" for their clothing

**Comment by Zireael07** [comment]
> but the merchant's name actually already appears in the header (bottom left of the top pane)

Didn't notice, thanks for pointing that out.

**Comment by ZeroInternalReflection** [comment]
I think I've captured all of the denial messages in the latest commit.  I rephrased "will never sell this" to "Won't sell their own equipment" since all of the checks for that seemed to be some variety of 'the NPC is currently wearing/using this'.  I'm not entirely sure about how the TRADER_KEEP flag is supposed to be used, but since I don't see anywhere that it _is_ used, I'm leaving it lumped in.

Testing:

Hub01_refusal messages

Smokes_refusal_messages_on_small_screen

Smokes_refusal_messages_on_large_screen

I'm not entirely happy with the little gap in highlighting on large screens, but I'm inclined to leave it for now, then hit everything on the selected line with a hilite_string() if/when #61447 gets merged.
Similarly, the trade UI header also gets quite cramped at small screen sizes.  I think that'll be a separate PR.

File: src/inventory_ui.cpp [hunk 1]
```diff
@@ -3014,6 +3016,18 @@ void inventory_multiselector::toggle_entries( int &count, const toggle_mode mode
...
+            const std::string denial = preset.get_denial( selected.front()->any_item() );
+            if( !denial.empty() ) {
+                const std::string assembled = selected.front()->any_item().get_item()->display_name() + ":\n"
+                                              + colorize( denial, c_red );
+                query_popup()
+                .message( "%s", assembled )
+                .option( "QUIT" )
+                .query();
+            }
+        }
```
Comment (andrei8l) [inline src/inventory_ui.cpp:3030]: The path `selected.empty()` to `selected.front()->...` leads to a segfault.

**Comment by Maleclypse** [comment]
@ZeroInternalReflection this needs a rebase and looks like a reviewer suggested a change to avoid a segfault?  Are you available to make those fixes?

**Comment by ZeroInternalReflection** [comment]
> @ZeroInternalReflection this needs a rebase and looks like a reviewer suggested a change to avoid a segfault? Are you available to make those fixes?

I've done the rebase and tweaked things to avoid the segfault, but for some reason the denial popup on selection no longer works.  I should have time to track that down and recheck the other changes later this week.

**Comment by ZeroInternalReflection** [comment]
> > @ZeroInternalReflection this needs a rebase and looks like a reviewer suggested a change to avoid a segfault? Are you available to make those fixes?
> 
> I've done the rebase and tweaked things to avoid the segfault, but for some reason the denial popup on selection no longer works. I should have time to track that down and recheck the other changes later this week.

Well, "later this week" sure didn't happen.  I got it working, but didn't get around to testing everything out (and getting astyle working again) until now.
All of the images and videos in the opening post have been updated to the current version.

The denial message is now highlighted, though on wider screens there can still be a gap in the highlighting, which is a bit annoying.

I'm still not sure exactly what related code change resulted in the popup breaking, but moving to highlighted_entry rather than selected seems to work.  At least, I haven't been able to segfault it _yet_.

**Pull request merged**